File: dotnet/CardsProject/Services/Cards/CardSetService.cs
using Microsoft.EntityFrameworkCore;
using System.ComponentModel.DataAnnotations;
using CardsProject.Controllers.Cards;
using CardsProject.Domain.Cards;
using CardsProject.Infrastructure;

namespace CardsProject.Services.Cards;

public class CardSetService
{
    private readonly AppDbContext _db;

    public CardSetService(AppDbContext db) => _db = db;

    private static void ValidateEntity(object entity)
    {
        var ctx = new ValidationContext(entity);
        Validator.ValidateObject(entity, ctx, validateAllProperties: true);
    }

    public async Task<List<CardSet>> GetAllAsync()
        => await _db.CardSets.AsNoTracking().ToListAsync();

    public async Task<CardSet?> GetByIdAsync(int id)
        => await _db.CardSets.FindAsync(id);

    public async Task<CardSet> CreateAsync(CardSetDto dto)
    {
        var entity = new CardSet();
        if (dto.Name is not null) entity.Name = dto.Name;
        if (dto.Code is not null) entity.Code = dto.Code;
        if (dto.ReleaseDate is not null) entity.ReleaseDate = dto.ReleaseDate.Value;
        if (dto.RotationDate is not null) entity.RotationDate = dto.RotationDate.Value;
        if (dto.SetType is not null && Enum.TryParse<CardSetSetTypeType>(dto.SetType, out var setTypeVal)) entity.SetType = setTypeVal;
        if (dto.TotalCards is not null) entity.TotalCards = dto.TotalCards.Value;
        if (dto.IsRotated is not null) entity.IsRotated = dto.IsRotated.Value;
        if (dto.Description is not null) entity.Description = dto.Description;
        if (dto.LogoUrl is not null) entity.LogoUrl = dto.LogoUrl;
        Validate(entity);
        ValidateEntity(entity);
        _db.CardSets.Add(entity);
        await _db.SaveChangesAsync();
        return entity;
    }

    public async Task<CardSet?> UpdateAsync(int id, CardSetDto dto)
    {
        var entity = await _db.CardSets.FindAsync(id);
        if (entity is null) return null;
        if (dto.Name is not null) entity.Name = dto.Name;
        if (dto.Code is not null) entity.Code = dto.Code;
        if (dto.ReleaseDate is not null) entity.ReleaseDate = dto.ReleaseDate.Value;
        if (dto.RotationDate is not null) entity.RotationDate = dto.RotationDate.Value;
        if (dto.SetType is not null && Enum.TryParse<CardSetSetTypeType>(dto.SetType, out var setTypeVal)) entity.SetType = setTypeVal;
        if (dto.TotalCards is not null) entity.TotalCards = dto.TotalCards.Value;
        if (dto.IsRotated is not null) entity.IsRotated = dto.IsRotated.Value;
        if (dto.Description is not null) entity.Description = dto.Description;
        if (dto.LogoUrl is not null) entity.LogoUrl = dto.LogoUrl;
        Validate(entity);
        ValidateEntity(entity);
        await _db.SaveChangesAsync();
        return entity;
    }

    public async Task<bool> DeleteAsync(int id)
    {
        var entity = await _db.CardSets.FindAsync(id);
        if (entity is null) return false;
        _db.CardSets.Remove(entity);
        await _db.SaveChangesAsync();
        return true;
    }

    public async System.Threading.Tasks.Task<bool> IsLegalInStandardAsync(int id)
    {
        var entity = await _db.CardSets.FindAsync(id);
        if (entity is null) throw new KeyNotFoundException("CardSet not found: " + id);
        var result = entity.IsLegalInStandard();
        await _db.SaveChangesAsync();
        return result;
    }
    public async System.Threading.Tasks.Task<bool> IsLegalInFormatAsync(int id, string format)
    {
        var entity = await _db.CardSets.FindAsync(id);
        if (entity is null) throw new KeyNotFoundException("CardSet not found: " + id);
        var result = entity.IsLegalInFormat(format);
        await _db.SaveChangesAsync();
        return result;
    }
    public async System.Threading.Tasks.Task<int> CardCountByRarityAsync(int id, string rarity)
    {
        var entity = await _db.CardSets.FindAsync(id);
        if (entity is null) throw new KeyNotFoundException("CardSet not found: " + id);
        var result = entity.CardCountByRarity(rarity);
        await _db.SaveChangesAsync();
        return result;
    }
    public async System.Threading.Tasks.Task<bool> RotateOutAsync(int id)
    {
        var entity = await _db.CardSets.FindAsync(id);
        if (entity is null) throw new KeyNotFoundException("CardSet not found: " + id);
        entity.RotateOut();
        await _db.SaveChangesAsync();
        return true;
    }
    public void Validate(CardSet entity)
    {
        if (entity.RotationDate != null && !((entity.RotationDate == null || (entity.ReleaseDate != null && entity.RotationDate > entity.ReleaseDate)))) throw new InvalidOperationException("Rotation date must be after release date");
        if (entity.IsRotated == true && entity.RotationDate == null) throw new InvalidOperationException("Rotated set must have a rotation date");
    }
}

File: spring/src/main/java/cardsproject/service/cards/CardSetService.java
package cardsproject.service.cards;

import cardsproject.domain.cards.CardSet;
import cardsproject.repository.cards.CardSetRepository;
import org.springframework.stereotype.Service;
import java.util.List;
import java.util.Optional;
import cardsproject.domain.cards.CardSetSetTypeType;

@Service
public class CardSetService {

    private final CardSetRepository repository;

    public CardSetService(CardSetRepository repository) {
        this.repository = repository;
    }

    public List<CardSet> findAll() {
        return repository.findAll();
    }

    public List<CardSet> search(String q) {
        if (q == null || q.isBlank()) return repository.findAll();
        return repository.findAll().stream()
            .filter(e -> (e.getName() != null && e.getName().toLowerCase().contains(q.toLowerCase())) || (e.getCode() != null && e.getCode().toLowerCase().contains(q.toLowerCase())))
            .collect(java.util.stream.Collectors.toList());
    }

    public Optional<CardSet> findById(Long id) {
        return repository.findById(id);
    }

    public CardSet save(CardSet entity) {
        validate(entity);
        return repository.save(entity);
    }

    public void delete(Long id) {
        repository.deleteById(id);
    }

    public void applyPatch(CardSet entity, java.util.Map<String, Object> patch) {
        if (patch.containsKey("name") && patch.get("name") != null) entity.setName(patch.get("name").toString());
        if (patch.containsKey("code") && patch.get("code") != null) entity.setCode(patch.get("code").toString());
        if (patch.containsKey("releaseDate") && patch.get("releaseDate") != null) entity.setReleaseDate(java.time.LocalDate.parse(patch.get("releaseDate").toString()));
        if (patch.containsKey("rotationDate") && patch.get("rotationDate") != null) entity.setRotationDate(java.time.LocalDate.parse(patch.get("rotationDate").toString()));
        if (patch.containsKey("setType")) entity.setSetType(CardSetSetTypeType.valueOf(patch.get("setType").toString()));
        if (patch.containsKey("totalCards") && patch.get("totalCards") != null) entity.setTotalCards(Integer.valueOf(patch.get("totalCards").toString()));
        if (patch.containsKey("isRotated") && patch.get("isRotated") != null) entity.setIsRotated(Boolean.valueOf(patch.get("isRotated").toString()));
        if (patch.containsKey("description") && patch.get("description") != null) entity.setDescription(patch.get("description").toString());
        if (patch.containsKey("logoUrl") && patch.get("logoUrl") != null) entity.setLogoUrl(patch.get("logoUrl").toString());
    }
    private void validate(CardSet entity) {
        if (entity.getRotationDate() != null && !((entity.getRotationDate() == null || (entity.getReleaseDate() != null && entity.getRotationDate().isAfter(entity.getReleaseDate()))))) throw new IllegalStateException("Rotation date must be after release date");
        if (Boolean.TRUE.equals(entity.getIsRotated()) && entity.getRotationDate() == null) throw new IllegalStateException("Rotated set must have a rotation date");
    }

    public Boolean isLegalInStandard(Long id) {
        CardSet entity = repository.findById(id)
            .orElseThrow(() -> new RuntimeException("CardSet not found: " + id));
        Boolean result = entity.isLegalInStandard();
        repository.save(entity);
        return result;
    }

    public Boolean isLegalInFormat(Long id, String format) {
        CardSet entity = repository.findById(id)
            .orElseThrow(() -> new RuntimeException("CardSet not found: " + id));
        Boolean result = entity.isLegalInFormat(format);
        repository.save(entity);
        return result;
    }

    public Integer cardCountByRarity(Long id, String rarity) {
        CardSet entity = repository.findById(id)
            .orElseThrow(() -> new RuntimeException("CardSet not found: " + id));
        Integer result = entity.cardCountByRarity(rarity);
        repository.save(entity);
        return result;
    }

    public void rotateOut(Long id) {
        CardSet entity = repository.findById(id)
            .orElseThrow(() -> new RuntimeException("CardSet not found: " + id));
        entity.rotateOut();
        repository.save(entity);
    }
}
